### DISCUSSION TITLE
issue: register pattern is set to undefined when valueAsNumber or valueAsDate is true

### DISCUSSION THREAD

**Discussion by FjVillar** [react-hook-form/react-hook-form#11493] [Q&A]
### Version Number

\ >= 7.43.0 

### Codesandbox/Expo snack

https://codesandbox.io/p/sandbox/dark-flower-xq5l5t?file=%2Fsrc%2FApp.tsx%3A56%2C14

### Steps to reproduce

First of all, thanks to all the contributors and maintainers of this library which has been very helpful to me these last years. You are doing an amazing work.

I think this could be an issue or maybe I was relying on a bug. I have created a sandbox to reproduce the error. I think this could be the [commit ](https://github.com/react-hook-form/react-hook-form/commit/17d472df6ed359dfdb2e119d5af14c955a833164)that is introducing this possible typing error.

After upgrading I found an error in my codebase and I started investigating a bit. I think this could be an unintended typing error or I'm doing a misuse of the `pattern` when the value is a number. What I wanted to do with this pattern is to alert the user when introducing letters like the `e` for exponential values which native HTML input allows even if you set the `type` attribute to `number`.

I don't want to allow my users to write the letter `E` or `e` for exponential values. 

Steps to reproduce: 

1. Use the `register` function to register an input that has `valueAsNumber` to true and try to add a `pattern` like in the code sandbox example

If this is not a bug and I'm doing something wrong, I would like to know how to achieve what I said with the current typings of the register function.

### Expected behaviour

It shouldn't type the `pattern` as `undefined` when valueAsNumber is `true`.

### What browsers are you seeing the problem on?

Chrome

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by Moshyfawn** [comment]
Hi! Thank you for your kind words 💜

Your reproduction link leads to a 404 page.

**Comment by FjVillar** [comment]
@Moshyfawn you are right, sorry. It was private. Edited original post. You should be able to see it now: https://codesandbox.io/p/sandbox/dark-flower-xq5l5t?file=%2Fsrc%2FApp.tsx%3A56%2C14
